File: electron/ipc/vault.ts
import { ipcMain, dialog } from 'electron'
import { promises as fs } from 'fs'
import * as path from 'path'
import { store } from '../store'
import { startWatcher, stopWatcher, markSelfWrite } from '../watcher'
import { buildIndex, updateFile } from '../mcp/search-index'

type VaultNode =
  | { kind: 'file'; path: string; name: string; mtime: number }
  | { kind: 'folder'; path: string; name: string; children: VaultNode[] }

async function readFolderOrder(dirPath: string): Promise<string[] | null> {
  const orderFile = path.join(dirPath, '.marrow-order.json')
  try {
    const raw = await fs.readFile(orderFile, 'utf-8')
    const parsed = JSON.parse(raw)
    if (parsed.version === 1 && Array.isArray(parsed.order)) return parsed.order as string[]
  } catch { /* no order file or invalid */ }
  return null
}

async function buildTree(dirPath: string): Promise<VaultNode[]> {
  let entries: import('fs').Dirent[]
  try {
    entries = await fs.readdir(dirPath, { withFileTypes: true })
  } catch {
    return []
  }

  const nodes: VaultNode[] = []

  for (const entry of entries) {
    // Skip dotfiles/dotfolders (includes .marrow-order.json)
    if (entry.name.startsWith('.')) continue

    const fullPath = path.join(dirPath, entry.name)

    if (entry.isDirectory()) {
      // Skip system/hidden directories (e.g. _attachments)
      if (entry.name.startsWith('_')) continue
      const children = await buildTree(fullPath)
      nodes.push({ kind: 'folder', path: fullPath, name: entry.name, children })
    } else if (entry.isFile() && entry.name.endsWith('.md')) {
      const stat = await fs.stat(fullPath)
      nodes.push({ kind: 'file', path: fullPath, name: entry.name, mtime: stat.mtimeMs })
    }
  }

  const orderList = await readFolderOrder(dirPath)

  if (orderList) {
    const byName = new Map(nodes.map(n => [n.name, n]))
    const ordered: VaultNode[] = []
    for (const name of orderList) {
      const node = byName.get(name)
      if (node) { ordered.push(node); byName.delete(name) }
    }
    // Nodes on disk but not in order file → sort by name, append after
    const remaining = Array.from(byName.values()).sort((a, b) => a.name.localeCompare(b.name))
    ordered.push(...remaining)
    return ordered
  }

  // Default sort: folders first (alpha), then files (mtime desc); CLAUDE.md pinned
  nodes.sort((a, b) => {
    if (a.kind !== b.kind) return a.kind === 'folder' ? -1 : 1
    if (a.kind === 'file' && b.kind === 'file') {
      if (a.name === 'CLAUDE.md') return -1
      if (b.name === 'CLAUDE.md') return 1
      return b.mtime - a.mtime
    }
    return a.name.localeCompare(b.name)
  })

  return nodes
}

export function registerVaultIPC() {
  ipcMain.handle('vault:pickVaultFolder', async () => {
    const result = await dialog.showOpenDialog({
      title: 'Choose Vault Folder',
      properties: ['openDirectory', 'createDirectory'],
    })
    if (result.canceled || result.filePaths.length === 0) return null
    return result.filePaths[0]
  })

  ipcMain.handle('vault:pickFile', async () => {
    const result = await dialog.showOpenDialog({
      title: 'Open File',
      properties: ['openFile'],
      filters: [
        { name: 'Markdown', extensions: ['md', 'markdown'] },
        { name: 'Text', extensions: ['txt'] },
        { name: 'All Files', extensions: ['*'] },
      ],
    })
    if (result.canceled || result.filePaths.length === 0) return null
    return result.filePaths[0]
  })

  ipcMain.handle('vault:getVaultPath', () => {
    return store.get('vaultPath') || null
  })

  ipcMain.handle('vault:setVaultPath', (_event, vaultPath: string) => {
    store.set('vaultPath', vaultPath)
    if (vaultPath) {
      startWatcher(vaultPath)
      buildIndex(vaultPath).catch(e =>
        console.error('[vault] Failed to rebuild search index:', e)
      )
    } else {
      stopWatcher()
    }
  })

  ipcMain.handle('vault:listTree', async () => {
    const vaultPath = store.get('vaultPath')
    if (!vaultPath) return []
    return buildTree(vaultPath)
  })

  ipcMain.handle('vault:readFile', async (_event, absPath: string) => {
    return fs.readFile(absPath, 'utf-8')
  })

  ipcMain.handle('vault:writeFile', async (_event, absPath: string, content: string) => {
    await fs.mkdir(path.dirname(absPath), { recursive: true })
    markSelfWrite(absPath)
    await fs.writeFile(absPath, content, 'utf-8')
  })

  ipcMain.handle('vault:createFile', async (_event, parentDir: string, name: string) => {
    const fileName = name.endsWith('.md') ? name : `${name}.md`
    const fullPath = path.join(parentDir, fileName)
    await fs.writeFile(fullPath, '', 'utf-8')
    return fullPath
  })

  ipcMain.handle('vault:createFolder', async (_event, parentDir: string, name: string) => {
    const fullPath = path.join(parentDir, name)
    await fs.mkdir(fullPath, { recursive: true })
    return fullPath
  })

  ipcMain.handle('vault:rename', async (_event, oldPath: string, newName: string) => {
    const newPath = path.join(path.dirname(oldPath), newName)
    await fs.rename(oldPath, newPath)
    return newPath
  })

  ipcMain.handle('vault:delete', async (_event, absPath: string) => {
    const { shell } = await import('electron')
    await shell.trashItem(absPath)
  })

  ipcMain.handle('vault:exists', async (_event, absPath: string) => {
    try {
      await fs.access(absPath)
      return true
    } catch {
      return false
    }
  })

  ipcMain.handle('vault:isInsideVault', (_event, absPath: string) => {
    const vaultPath = store.get('vaultPath')
    if (!vaultPath) return false
    return path.resolve(absPath).startsWith(path.resolve(vaultPath))
  })

  ipcMain.handle('vault:move', async (_event, oldPath: string, newParentDir: string) => {
    const name = path.basename(oldPath)
    const newPath = path.join(newParentDir, name)
    try {
      await fs.access(newPath)
      return { conflict: true, newPath }
    } catch { /* no conflict */ }
    await fs.rename(oldPath, newPath)
    return { conflict: false, newPath }
  })

  ipcMain.handle('vault:setFolderOrder', async (_event, folderPath: string, order: string[]) => {
    const orderFile = path.join(folderPath, '.marrow-order.json')
    await fs.writeFile(orderFile, JSON.stringify({ version: 1, order }, null, 2), 'utf-8')
  })

  ipcMain.handle('vault:getFolderOrder', async (_event, folderPath: string) => {
    const orderFile = path.join(folderPath, '.marrow-order.json')
    try {
      const raw = await fs.readFile(orderFile, 'utf-8')
      const parsed = JSON.parse(raw)
      if (parsed.version === 1 && Array.isArray(parsed.order)) return parsed.order as string[]
    } catch { /* no order file */ }
    return null
  })

  ipcMain.handle('vault:fileExists', async (_event, relPath: string) => {
    const vaultPath = store.get('vaultPath')
    if (!vaultPath) return false
    const abs = path.resolve(vaultPath, relPath)
    // Safety: must stay within vault
    if (!abs.startsWith(path.resolve(vaultPath))) return false
    try {
      await fs.access(abs)
      return true
    } catch {
      return false
    }
  })

  ipcMain.handle('vault:writeBinary', async (_event, relPath: string, base64: string) => {
    const vaultPath = store.get('vaultPath')
    if (!vaultPath) throw new Error('No vault path set')
    const abs = path.resolve(vaultPath, relPath)
    // Safety: must stay within vault
    if (!abs.startsWith(path.resolve(vaultPath))) throw new Error('Path escapes vault')
    const buf = Buffer.from(base64, 'base64')
    await fs.mkdir(path.dirname(abs), { recursive: true })
    await fs.writeFile(abs, buf)
    return { ok: true, relPath }
  })

  /**
   * Copy an external file from the user's filesystem into the vault.
   * `targetFolderAbs` is either an absolute path inside the vault (e.g. when
   * dropped on a folder in the tree) or a vault-relative path like 'inbox'.
   *
   * conflict:
   *   - 'fail'       — error out with reason: 'conflict' and the existing path
   *   - 'replace'    — overwrite the existing file
   *   - 'keep-both'  — append " (1)", " (2)", … until the name is free
   */
  ipcMain.handle('vault:importFile', async (
    _event,
    srcAbsPath: string,
    targetFolder: string,
    conflict: 'fail' | 'replace' | 'keep-both' = 'keep-both'
  ): Promise<
    | { ok: true; path: string; renamedFrom?: string }
    | { ok: false; reason: 'conflict' | 'outside-vault' | 'no-vault' | 'error'; message?: string; existing?: string }
  > => {
    try {
      const vaultPath = store.get('vaultPath')
      if (!vaultPath) return { ok: false, reason: 'no-vault' }
      const vaultAbs = path.resolve(vaultPath)

      // Resolve target dir: absolute path or vault-relative.
      const targetAbs = path.isAbsolute(targetFolder)
        ? path.resolve(targetFolder)
        : path.resolve(vaultAbs, targetFolder)

      if (!targetAbs.startsWith(vaultAbs)) {
        return { ok: false, reason: 'outside-vault' }
      }

      await fs.mkdir(targetAbs, { recursive: true })

      const originalName = path.basename(srcAbsPath)
      let destAbs = path.join(targetAbs, originalName)
      let renamedFrom: string | undefined

      const exists = async (p: string) => {
        try { await fs.access(p); return true } catch { return false }
      }

      if (await exists(destAbs)) {
        if (conflict === 'fail') {
          return { ok: false, reason: 'conflict', existing: destAbs }
        }
        if (conflict === 'keep-both') {
          renamedFrom = originalName
          const ext = path.extname(originalName)
          const stem = ext ? originalName.slice(0, -ext.length) : originalName
          for (let i = 1; i < 1000; i++) {
            const candidate = path.join(targetAbs, `${stem} (${i})${ext}`)
            if (!(await exists(candidate))) {
              destAbs = candidate
              break
            }
          }
        }
        // 'replace' falls through to copyFile and overwrites
      }

      // Mark as self-write so the file watcher doesn't pop a "changed externally" dialog.
      markSelfWrite(destAbs)
      await fs.copyFile(srcAbsPath, destAbs)

      return renamedFrom
        ? { ok: true, path: destAbs, renamedFrom }
        : { ok: true, path: destAbs }
    } catch (err) {
      return { ok: false, reason: 'error', message: err instanceof Error ? err.message : String(err) }
    }
  })
}

File: src-tauri/src/commands/vault.rs
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use tauri::AppHandle;
use tauri_plugin_store::StoreExt;

use crate::error::AppError;

const STORE_FILE: &str = "settings.json";

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "kind")]
pub enum VaultNode {
    #[serde(rename = "file")]
    File {
        path: String,
        name: String,
        mtime: f64,
    },
    #[serde(rename = "folder")]
    Folder {
        path: String,
        name: String,
        children: Vec<VaultNode>,
    },
}

#[derive(Debug, Serialize)]
pub struct MoveResult {
    pub conflict: bool,
    #[serde(rename = "newPath")]
    pub new_path: String,
}

#[derive(Debug, Serialize)]
pub struct WriteBinaryResult {
    pub ok: bool,
    #[serde(rename = "relPath")]
    pub rel_path: String,
}

fn get_vault_path(app: &AppHandle) -> Result<Option<String>, AppError> {
    let store = app.store(STORE_FILE)?;
    match store.get("vaultPath") {
        Some(serde_json::Value::String(s)) => Ok(Some(s)),
        _ => Ok(None),
    }
}

fn validate_inside(base: &Path, target: &Path) -> Result<PathBuf, AppError> {
    let canon_base = std::fs::canonicalize(base).map_err(|e| AppError::from(e.to_string()))?;
    let canon_target = std::fs::canonicalize(target).map_err(|e| AppError::from(e.to_string()))?;
    if !canon_target.starts_with(&canon_base) {
        return Err(AppError::from("Path is outside vault"));
    }
    Ok(canon_target)
}

fn validate_parent_inside(base: &Path, target: &Path) -> Result<(), AppError> {
    let canon_base = std::fs::canonicalize(base).map_err(|e| AppError::from(e.to_string()))?;
    if let Some(parent) = target.parent() {
        if parent.exists() {
            let canon_parent =
                std::fs::canonicalize(parent).map_err(|e| AppError::from(e.to_string()))?;
            if !canon_parent.starts_with(&canon_base) {
                return Err(AppError::from("Path is outside vault"));
            }
        }
    }
    Ok(())
}

#[derive(Debug, Serialize, Deserialize)]
struct FolderOrder {
    version: u32,
    order: Vec<String>,
}

async fn read_folder_order(dir: &Path) -> Option<Vec<String>> {
    let order_file = dir.join(".marrow-order.json");
    let raw = tokio::fs::read_to_string(&order_file).await.ok()?;
    let parsed: FolderOrder = serde_json::from_str(&raw).ok()?;
    if parsed.version == 1 {
        Some(parsed.order)
    } else {
        None
    }
}

async fn build_tree(dir: &Path) -> Vec<VaultNode> {
    let mut entries = match tokio::fs::read_dir(dir).await {
        Ok(e) => e,
        Err(_) => return vec![],
    };

    let mut nodes = Vec::new();

    while let Ok(Some(entry)) = entries.next_entry().await {
        let name = entry.file_name().to_string_lossy().to_string();
        if name.starts_with('.') {
            continue;
        }

        let full_path = entry.path();
        let ft = match entry.file_type().await {
            Ok(ft) => ft,
            Err(_) => continue,
        };

        if ft.is_dir() {
            if name.starts_with('_') {
                continue;
            }
            let children = Box::pin(build_tree(&full_path)).await;
            nodes.push(VaultNode::Folder {
                path: full_path.to_string_lossy().to_string(),
                name,
                children,
            });
        } else if ft.is_file() && name.ends_with(".md") {
            let mtime = tokio::fs::metadata(&full_path)
                .await
                .map(|m| {
                    m.modified()
                        .ok()
                        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                        .map(|d| d.as_secs_f64() * 1000.0)
                        .unwrap_or(0.0)
                })
                .unwrap_or(0.0);
            nodes.push(VaultNode::File {
                path: full_path.to_string_lossy().to_string(),
                name,
                mtime,
            });
        }
    }

    let order_list = read_folder_order(dir).await;

    if let Some(order) = order_list {
        let mut by_name: std::collections::HashMap<String, VaultNode> =
            nodes.into_iter().map(|n| (node_name(&n).to_string(), n)).collect();
        let mut ordered = Vec::new();
        for name in &order {
            if let Some(node) = by_name.remove(name) {
                ordered.push(node);
            }
        }
        let mut remaining: Vec<VaultNode> = by_name.into_values().collect();
        remaining.sort_by(|a, b| node_name(a).cmp(node_name(b)));
        ordered.extend(remaining);
        ordered
    } else {
        nodes.sort_by(|a, b| {
            let a_is_folder = matches!(a, VaultNode::Folder { .. });
            let b_is_folder = matches!(b, VaultNode::Folder { .. });
            if a_is_folder != b_is_folder {
                return if a_is_folder {
                    std::cmp::Ordering::Less
                } else {
                    std::cmp::Ordering::Greater
                };
            }
            if !a_is_folder {
                let a_name = node_name(a);
                let b_name = node_name(b);
                if a_name == "CLAUDE.md" {
                    return std::cmp::Ordering::Less;
                }
                if b_name == "CLAUDE.md" {
                    return std::cmp::Ordering::Greater;
                }
                let a_mtime = node_mtime(a);
                let b_mtime = node_mtime(b);
                return b_mtime
                    .partial_cmp(&a_mtime)
                    .unwrap_or(std::cmp::Ordering::Equal);
            }
            node_name(a).cmp(node_name(b))
        });
        nodes
    }
}

fn node_name(n: &VaultNode) -> &str {
    match n {
        VaultNode::File { name, .. } => name,
        VaultNode::Folder { name, .. } => name,
    }
}

fn node_mtime(n: &VaultNode) -> f64 {
    match n {
        VaultNode::File { mtime, .. } => *mtime,
        VaultNode::Folder { .. } => 0.0,
    }
}

// ── Commands ─────────────────────────────────────────────────────────────────

#[tauri::command]
pub fn pick_vault_folder(app: AppHandle) -> Result<Option<String>, AppError> {
    use tauri_plugin_dialog::DialogExt;
    log::info!("pick_vault_folder: opening dialog");
    let folder = app.dialog().file().set_title("Choose Vault Folder").blocking_pick_folder();
    log::info!("pick_vault_folder: result = {:?}", folder.as_ref().map(|p| p.to_string()));
    Ok(folder.map(|p| p.to_string()))
}

#[tauri::command]
pub fn pick_file(app: AppHandle) -> Result<Option<String>, AppError> {
    use tauri_plugin_dialog::DialogExt;
    let file = app
        .dialog()
        .file()
        .set_title("Open File")
        .add_filter("Markdown", &["md", "markdown"])
        .add_filter("Text", &["txt"])
        .add_filter("All Files", &["*"])
        .blocking_pick_file();
    Ok(file.map(|p| p.to_string()))
}

#[tauri::command]
pub async fn vault_get_path(app: AppHandle) -> Result<Option<String>, AppError> {
    get_vault_path(&app)
}

#[tauri::command]
pub async fn vault_set_path(app: AppHandle, path: String) -> Result<(), AppError> {
    let store = app.store(STORE_FILE)?;
    store.set("vaultPath", serde_json::Value::String(path));
    Ok(())
}

#[tauri::command]
pub async fn list_tree(app: AppHandle) -> Result<Vec<VaultNode>, AppError> {
    let vault_path = match get_vault_path(&app)? {
        Some(p) => p,
        None => return Ok(vec![]),
    };
    Ok(build_tree(Path::new(&vault_path)).await)
}

#[tauri::command]
pub async fn read_file(path: String) -> Result<String, AppError> {
    tokio::fs::read_to_string(&path)
        .await
        .map_err(|e| AppError::from(e.to_string()))
}

#[tauri::command]
pub async fn write_file(path: String, content: String) -> Result<(), AppError> {
    if let Some(parent) = Path::new(&path).parent() {
        tokio::fs::create_dir_all(parent)
            .await
            .map_err(|e| AppError::from(e.to_string()))?;
    }
    tokio::fs::write(&path, &content)
        .await
        .map_err(|e| AppError::from(e.to_string()))
}

#[tauri::command]
pub async fn create_file(dir: String, name: String) -> Result<String, AppError> {
    let file_name = if name.ends_with(".md") {
        name
    } else {
        format!("{}.md", name)
    };
    let full_path = Path::new(&dir).join(&file_name);
    tokio::fs::write(&full_path, "")
        .await
        .map_err(|e| AppError::from(e.to_string()))?;
    Ok(full_path.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn create_folder(dir: String, name: String) -> Result<String, AppError> {
    let full_path = Path::new(&dir).join(&name);
    tokio::fs::create_dir_all(&full_path)
        .await
        .map_err(|e| AppError::from(e.to_string()))?;
    Ok(full_path.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn rename_file(old_path: String, new_name: String) -> Result<String, AppError> {
    let old = Path::new(&old_path);
    let new_path = old
        .parent()
        .unwrap_or(Path::new(""))
        .join(&new_name);
    tokio::fs::rename(&old, &new_path)
        .await
        .map_err(|e| AppError::from(e.to_string()))?;
    Ok(new_path.to_string_lossy().to_string())
}

#[tauri::command]
pub async fn delete_file(path: String) -> Result<(), AppError> {
    let p = Path::new(&path);
    if p.is_dir() {
        tokio::fs::remove_dir_all(p)
            .await
            .map_err(|e| AppError::from(e.to_string()))?;
    } else {
        tokio::fs::remove_file(p)
            .await
            .map_err(|e| AppError::from(e.to_string()))?;
    }
    Ok(())
}

#[tauri::command]
pub async fn file_exists(path: String) -> Result<bool, AppError> {
    Ok(Path::new(&path).exists())
}

#[tauri::command]
pub async fn is_inside_vault(app: AppHandle, path: String) -> Result<bool, AppError> {
    let vault_path = match get_vault_path(&app)? {
        Some(p) => p,
        None => return Ok(false),
    };
    let vault_abs = std::fs::canonicalize(&vault_path).unwrap_or_else(|_| PathBuf::from(&vault_path));
    let target_abs = std::fs::canonicalize(&path).unwrap_or_else(|_| PathBuf::from(&path));
    Ok(target_abs.starts_with(&vault_abs))
}

#[tauri::command]
pub async fn move_file(old_path: String, new_parent_dir: String) -> Result<MoveResult, AppError> {
    let name = Path::new(&old_path)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();
    let new_path = Path::new(&new_parent_dir).join(&name);
    if new_path.exists() {
        return Ok(MoveResult {
            conflict: true,
            new_path: new_path.to_string_lossy().to_string(),
        });
    }
    tokio::fs::rename(&old_path, &new_path)
        .await
        .map_err(|e| AppError::from(e.to_string()))?;
    Ok(MoveResult {
        conflict: false,
        new_path: new_path.to_string_lossy().to_string(),
    })
}

#[tauri::command]
pub async fn set_folder_order(
    folder_path: String,
    order: Vec<String>,
) -> Result<(), AppError> {
    let order_file = Path::new(&folder_path).join(".marrow-order.json");
    let data = FolderOrder { version: 1, order };
    let json = serde_json::to_string_pretty(&data).map_err(|e| AppError::from(e.to_string()))?;
    tokio::fs::write(&order_file, json)
        .await
        .map_err(|e| AppError::from(e.to_string()))
}

#[tauri::command]
pub async fn get_folder_order(folder_path: String) -> Result<Option<Vec<String>>, AppError> {
    Ok(read_folder_order(Path::new(&folder_path)).await)
}

#[tauri::command]
pub async fn vault_file_exists(app: AppHandle, rel_path: String) -> Result<bool, AppError> {
    let vault_path = match get_vault_path(&app)? {
        Some(p) => p,
        None => return Ok(false),
    };
    let abs = Path::new(&vault_path).join(&rel_path);
    if let Ok(canon) = std::fs::canonicalize(&abs) {
        let vault_canon =
            std::fs::canonicalize(&vault_path).unwrap_or_else(|_| PathBuf::from(&vault_path));
        if !canon.starts_with(&vault_canon) {
            return Ok(false);
        }
    }
    Ok(abs.exists())
}

#[tauri::command]
pub async fn write_binary(
    app: AppHandle,
    rel_path: String,
    base64: String,
) -> Result<WriteBinaryResult, AppError> {
    use base64::Engine;
    let vault_path = get_vault_path(&app)?
        .ok_or_else(|| AppError::from("No vault path set"))?;
    let abs = Path::new(&vault_path).join(&rel_path);
    validate_parent_inside(Path::new(&vault_path), &abs)?;
    if let Some(parent) = abs.parent() {
        tokio::fs::create_dir_all(parent)
            .await
            .map_err(|e| AppError::from(e.to_string()))?;
    }
    let bytes = base64::engine::general_purpose::STANDARD
        .decode(&base64)
        .map_err(|e| AppError::from(e.to_string()))?;
    tokio::fs::write(&abs, bytes)
        .await
        .map_err(|e| AppError::from(e.to_string()))?;
    Ok(WriteBinaryResult {
        ok: true,
        rel_path,
    })
}
